azext_iot/tests/iothub/__init__.py

```diff
@@ -130,41 +139,74 @@ def __init__(self, test_scenario, add_data_contributor=True):
             test_tag=test_scenario
         )
 
-    def clean_up(self, device_ids: List[str] = None, config_ids: List[str] = None):
-        if device_ids:
-            device = device_ids.pop()
+    def _add_data_contributor(self, target_hub):
+        account = self.cmd("account show").get_output_in_json()
+        user = account["user"]
+
+        if user["name"] is None:
+            raise Exception("User not found")
+
+        tries = 0
+        while tries < MAX_RBAC_ASSIGNMENT_TRIES:
+            role_assignments = self.get_role_assignments(target_hub["id"], USER_ROLE)
+            role_assignment_principal_names = [assignment["principalName"] for assignment in role_assignments]
+            if user["name"] in role_assignment_principal_names:
+                break
+            # else assign IoT Hub Data Contributor role to current user and check again
             self.cmd(
-                "iot hub device-identity delete -d {} --login {}".format(
-                    device, self.connection_string
-                ),
-                checks=self.is_empty(),
+                'role assignment create --assignee "{}" --role "{}" --scope "{}"'.format(
+                    user["name"], USER_ROLE, target_hub["id"]
+                )
             )
+            sleep(10)
 
-            for device in device_ids:
+        if tries == MAX_RBAC_ASSIGNMENT_TRIES:
+            raise Exception(
+                "Reached max ({}) number of tries to assign RBAC role. Please re-run the test later "
+                "or with more max number of tries.".format(MAX_RBAC_ASSIGNMENT_TRIES)
+            )
+
+    def clean_up(self, device_ids: List[str] = None, config_ids: List[str] = None):
```

Comment:
you should not need to keep track of created devices and original devices during the test. Pick one and then use the list commands to get all devices and you can figure out the other half from there.